messaging/tck/src/main/java/org/eclipse/microprofile/reactive/messaging/tck/arquillian/TckArquillianDeployListener.java

```diff
@@ -0,0 +1,141 @@
+
+package org.eclipse.microprofile.reactive.messaging.tck.arquillian;
+
+import org.eclipse.microprofile.reactive.messaging.tck.ContainerController;
+import org.eclipse.microprofile.reactive.messaging.tck.MockedReceiver;
+import org.eclipse.microprofile.reactive.messaging.tck.MockedSender;
+import org.eclipse.microprofile.reactive.messaging.tck.ReceiveTimeoutException;
+import org.eclipse.microprofile.reactive.messaging.tck.TckMessagingManager;
+import org.eclipse.microprofile.reactive.messaging.tck.spi.TestEnvironment;
+import org.eclipse.microprofile.reactive.messaging.tck.mocks.MockPayload;
+import org.eclipse.microprofile.reactive.messaging.tck.mocks.SimpleMessage;
+import org.eclipse.microprofile.reactive.messaging.tck.spi.TckContainer;
+import org.jboss.arquillian.container.spi.client.container.DeployableContainer;
+import org.jboss.arquillian.container.spi.client.container.DeploymentException;
+import org.jboss.arquillian.container.spi.client.deployment.DeploymentDescription;
+import org.jboss.arquillian.container.spi.event.container.AfterUnDeploy;
+import org.jboss.arquillian.container.spi.event.container.BeforeDeploy;
+import org.jboss.arquillian.core.api.annotation.Observes;
+import org.jboss.arquillian.test.spi.TestClass;
+import org.jboss.shrinkwrap.api.Archive;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class TckArquillianDeployListener {
+
+  private final TckContainer tckContainer;
+
+  private final Map<Class<?>, List<DeploymentDescription>> activeDeployments = new ConcurrentHashMap<>();
+
+  /**
+   * I don't know how deployments work in arquillian, I assume I need to explicitly deploy everything?
+   */
+  private final Archive<?> frameworkArchive = ShrinkWrap.create(JavaArchive.class)
+      .addClass(MockedReceiver.class)
+      .addClass(MockedSender.class)
+      .addClass(ReceiveTimeoutException.class)
+      .addClass(TestEnvironment.class)
+      .addClass(MockPayload.class)
+      .addClass(SimpleMessage.class)
+      .addClass(TckMessagingManager.class)
+      .addClass(ContainerController.class);
+
+
+  public TckArquillianDeployListener() {
+    Iterator<TckContainer> containers = ServiceLoader.load(TckContainer.class).iterator();
+
+    if (containers.hasNext()) {
+      this.tckContainer = containers.next();
+    }
+    else {
+      throw new RuntimeException("No " + TckContainer.class.getName() +
+          " found. To run this TCK, you must provide an implementation of " +
+          TckContainer.class + " via the JDK service loader mechanism.");
+    }
+  }
+
+  public void onBeforeDeploy(@Observes BeforeDeploy beforeDeploy, TestClass testClass) throws DeploymentException {
+    Topics topics = testClass.getAnnotation(Topics.class);
```

Comment:
I would add all this logic of getting the annotation and if/else into a private method to make the onBeforeDeploy clear